File: app/server/utils/encrypt.js
const crypto = require('crypto');
const CryptoJS = require("crypto-js");
const bcrypt = require('bcrypt');
const fs = require('fs');
const moment = require('moment');
'use strict'

class Encrypt {
    static generateDigest = (payload) => {
        return crypto.createHash('sha256').update(payload, 'utf-8').digest('base64');
    };

    static generateSignature = (signingString) => {
        try {
            const privateKey = fs.readFileSync('./certs/bcr/bcr_sign_private_key.key');
            return crypto.createSign('sha256').update(signingString).sign(privateKey, 'base64');
        }
        catch (err) {
            console.log(err);
        } return "";
    };

    static generateSigningString = (method, endpoint, date, digest) => {
        return '(request-target): ' + method.toLowerCase() + ' ' + endpoint
            + '\n' + 'date: ' + date
            + '\n' + 'digest: ' + digest;
    };

    static buildSignatureHeaderWithKeyId = (method, endpoint, date, digest, keyId) => {
        const signingString = Encrypt.generateSigningString(method, endpoint, date, digest);
        const signature = Encrypt.generateSignature(signingString);
        return 'keyId="' + keyId + '\",algorithm=\"rsa-sha256\",headers=\"(request-target) date digest\",signature=\"' + signature + '"';
    };

    static buildBearerTokenHeader = (token) => {
        return 'Bearer ' + token;
    }

    static buildDateHeader = () => {
        return moment.utc().format('ddd, DD MMM yyyy HH:mm:ss') + ' GMT';
    };

    static buildDigestHeader = (payload) => {
        return 'SHA-256=' + Encrypt.generateDigest(payload);
    };

    static encryptData = (data) => {
        return CryptoJS.AES.encrypt(data, CryptoJS.enc.Utf8.parse(process.env.AES_SECRET), {
            mode: CryptoJS.mode.CBC,
            iv: CryptoJS.enc.Hex.parse(process.env.IV)
        }).toString();
    };

    static decryptData = (data) => {
        return CryptoJS.AES.decrypt(data, CryptoJS.enc.Utf8.parse(process.env.AES_SECRET), {
            mode: CryptoJS.mode.CBC,
            iv: CryptoJS.enc.Hex.parse(process.env.IV),
        }).toString(CryptoJS.enc.Utf8);
    }

    static hashData = (data) => {
        return bcrypt.hashSync(data, parseInt(process.env.SALT_ROUNDS));
    };
}

module.exports = {
    fs,
    crypto,
    Encrypt
};

File: app/src/main/java/licenta/allbank/utils/Encrypt.java
package licenta.allbank.utils;

public class Encrypt {
//    public static String generateDigest(String payload) throws NoSuchAlgorithmException {
////        echo -n "$payload" | openssl dgst -binary -sha256 | openssl base64
//        java.security.MessageDigest digest = java.security.MessageDigest.getInstance("SHA-256");
//        digest.reset();
//        digest.update(payload.getBytes(StandardCharsets.UTF_8));
//        return Base64.getEncoder().encodeToString(digest.digest());
//    }
//
//    public static String generateSignature(Context context, String signingString) throws NoSuchAlgorithmException, KeyStoreException, IOException, CertificateException, UnrecoverableEntryException, InvalidKeyException, SignatureException {
////        printf %s "$signingString" | openssl dgst -sha256 -sign "${certPath}example_eidas_client_signing.key" | openssl base64 -A
//        KeyStore keyStore = KeyStore.getInstance("BKS");
//        keyStore.load(context.getAssets().open("keystore_ing_signing_certificate.bks"), ServiceServer.getKeystoreSecretIng());
//
//        KeyStore.PrivateKeyEntry privateKeyEntry = (KeyStore.PrivateKeyEntry) keyStore.getEntry(
//                keyStore.aliases().nextElement(),
//                new KeyStore.PasswordProtection(ServiceServer.getKeystoreSecretIng())
//        );
//        PrivateKey privateKey = privateKeyEntry.getPrivateKey();
//
//        Signature signature = Signature.getInstance("SHA256withRSA");
//        signature.initSign(privateKey);
//        signature.update(signingString.getBytes());
//        byte[] stringSignature = signature.sign();
//
//        return Base64.getEncoder().encodeToString(stringSignature);
//    }
//
//    public static String generateSigningString(String method, String endpoint, String date, String digest) {
//        return "(request-target): " + method.toLowerCase() + " " + endpoint
//                + "\n" + "date: " + date
//                + "\n" + "digest: " + digest;
//    }

//    public static String buildSignatureHeader(String signature) {
//        return "keyId=\"5ca1ab1e-c0ca-c01a-cafe-154deadbea75\",algorithm=\"rsa-sha256\",headers=\"(request-target) date digest\",signature=\"" + signature + "\"";
//    }

    public static String buildBearerTokenHeader(String token) {
        return "Bearer " + token;
    }

//    public static String buildDateHeader(String date) {
//        return date + " GMT";
//    }
//
//    public static String buildDigestHeader(String digest) {
//        return "SHA-256=" + digest;
//    }
}